File: functionaltests/source/write_logs.py
"""Write some logs."""
from __future__ import print_function

import sys
import time
import os
import logging
logging.basicConfig(stream=sys.stderr, level=logging.DEBUG)

from mdk import start
mdk = start()

start = time.time()

def main():
    session = mdk.session()
    session.trace("DEBUG")
    category = sys.argv[1]
    results = set()
    expected = set(("hello {} {}".format(category, level), level.upper())
                   for level in ["critical", "error", "warn", "info", "debug"])

    def got_message(event):
        if time.time() - start > 60:
            print("Took more than 60 seconds, giving up.")
            os._exit(1)

        print((event.category, event.text))
        if event.category == category:
            results.add((event.text, event.level))
        if expected == results:
            print("Got expected messages!")
            mdk.stop()
            # XXX working around Quark/MDK issue where Python runtime isn't exiting:
            os._exit(0)
        else:
            print("No full results yet. Specifically got: " + repr(event))

    mdk._tracer.subscribe(got_message)
    time.sleep(1)
    for level in ["critical", "error", "warn", "info", "debug"]:
        f = getattr(session, level)
        f(category, "hello {} {}".format(category, level))


if __name__ == '__main__':
    main()

File: functionaltests/source/write_logs.js
"use strict";

var process = require("process");
var mdk = require("datawire_mdk").mdk;
var MDK = mdk.start();

var main = function() {
    setTimeout(function() {
        console.log("timeout");
        MDK.stop();
        process.exit(1);
    }, 30000);
    var session = MDK.session();
    session.trace("DEBUG");
    var category = process.argv[2];
    var messages = [];
    var sent_messages = ["hello critical " + category,
                         "hello debug " + category,
                         "hello error " + category,
                         "hello info " + category,
                         "hello warn " + category];
    MDK._tracer.subscribe(function (event) {
        if (event.category == category) {
            messages.push(event.text);
        }
        if (messages.length == 5) {
            MDK.stop();
            messages.sort();
            // Remind me never to write any Javascript ever again:
            if (JSON.stringify(messages) != JSON.stringify(sent_messages)) {
                console.log("unexpected responses: " + messages);
                process.exit(1);
            } else {
                console.log("got all messages");
                process.exit(0);
            }
        }
    });

    setTimeout(function() {
        session.critical(category, sent_messages[0]);
        session.debug(category, sent_messages[1]);
        session.error(category, sent_messages[2]);
        session.info(category, sent_messages[3]);
        session.warn(category, sent_messages[4]);
    }, 3000);
};

main();
